File: frontend/js/views/income_statement.js
// Income statement (P&L) generator — revenue → gross profit → EBIT → pre-tax →
// net income with margins, via /calc/income-statement.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
let LAST_DOC = null;

export async function renderIncomeStatement(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.incstmt.h1.title">// INCOME STATEMENT</span></h1>
        <p class="muted small" data-i18n="view.incstmt.hint.intro">
            The multi-step business P&L: revenue less cost of goods sold gives gross profit; less operating
            expenses gives operating income (EBIT); less interest gives pre-tax income; less income tax gives
            net income. It reports the margin at each level as a percentage of revenue. Drafting aid, not
            accounting advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.incstmt.h2.inputs">Statement inputs</h2>
            <form id="is-form" class="inline-form">
                <label><span data-i18n="view.incstmt.label.company">Company</span>
                    <input type="text" name="company_name" value="Acme Co" required></label>
                <label><span data-i18n="view.incstmt.label.period">Period</span>
                    <input type="text" name="period_label" value="FY2025" required></label>
                <label><span data-i18n="view.incstmt.label.revenue">Revenue ($)</span>
                    <input type="number" step="1000" min="0" name="revenue_usd" value="1000000" required></label>
                <label><span data-i18n="view.incstmt.label.cogs">Cost of goods sold ($)</span>
                    <input type="number" step="1000" min="0" name="cogs_usd" value="600000"></label>
                <label><span data-i18n="view.incstmt.label.opex">Operating expenses ($)</span>
                    <input type="number" step="1000" min="0" name="operating_expenses_usd" value="250000"></label>
                <label><span data-i18n="view.incstmt.label.interest">Interest expense ($)</span>
                    <input type="number" step="1000" min="0" name="interest_expense_usd" value="20000"></label>
                <label><span data-i18n="view.incstmt.label.tax">Tax rate (%)</span>
                    <input type="number" step="0.1" min="0" name="tax_rate_pct" value="21"></label>
                <label><span data-i18n="view.incstmt.label.date">Date</span>
                    <input type="date" name="date" value="2026-01-31" required></label>
            </form>
        </div>
        <div id="is-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#is-form');
    const num = (k) => Number(form.querySelector(`[name="${k}"]`).value) || 0;
    const generate = async () => {
        const body = {
            company_name: (form.querySelector('[name="company_name"]').value || '').trim(),
            period_label: (form.querySelector('[name="period_label"]').value || '').trim(),
            revenue_usd: num('revenue_usd'),
            cogs_usd: num('cogs_usd'),
            operating_expenses_usd: num('operating_expenses_usd'),
            interest_expense_usd: num('interest_expense_usd'),
            tax_rate_pct: num('tax_rate_pct'),
            date: form.querySelector('[name="date"]').value,
        };
        try {
            const doc = await api.calcIncomeStatement(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.incstmt.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase(), ''];
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const niCls = doc.net_income_usd >= 0 ? 'pos' : 'neg';
    const el = mount.querySelector('#is-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card ${niCls}"><div class="label" data-i18n="view.incstmt.card.net">Net income</div>
                    <div class="value">${money(doc.net_income_usd)} (${pct(doc.net_margin_pct)})</div></div>
                <div class="card"><div class="label" data-i18n="view.incstmt.card.gross">Gross profit</div>
                    <div class="value">${money(doc.gross_profit_usd)} (${pct(doc.gross_margin_pct)})</div></div>
                <div class="card"><div class="label" data-i18n="view.incstmt.card.ebit">Operating income</div>
                    <div class="value">${money(doc.operating_income_usd)} (${pct(doc.operating_margin_pct)})</div></div>
                <div class="card"><div class="label" data-i18n="view.incstmt.card.pretax">Pre-tax income</div>
                    <div class="value">${money(doc.pretax_income_usd)} (${pct(doc.pretax_margin_pct)})</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="is-copy" type="button" data-i18n="view.incstmt.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="is-download" type="button" data-i18n="view.incstmt.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#is-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.incstmt.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.incstmt.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#is-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'income-statement.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/income_statement.rs
//! Income statement (profit & loss) — the standard multi-step business P&L
//! cascade: revenue less cost of goods sold gives gross profit; less operating
//! expenses gives operating income (EBIT); less interest gives pre-tax income;
//! less income tax gives net income. It also reports the margin at each level as
//! a percentage of revenue. Distinct from the trading P&L modules and from the
//! financial-ratios module; this assembles the statement itself. Drafting aid,
//! not accounting advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct IncomeStatementInput {
    pub company_name: String,
    pub period_label: String,
    pub revenue_usd: f64,
    #[serde(default)]
    pub cogs_usd: f64,
    #[serde(default)]
    pub operating_expenses_usd: f64,
    #[serde(default)]
    pub interest_expense_usd: f64,
    /// Income tax rate on pre-tax income, percent.
    #[serde(default)]
    pub tax_rate_pct: f64,
    pub date: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct IncomeStatement {
    pub title: String,
    /// Revenue − COGS.
    pub gross_profit_usd: f64,
    /// Gross profit − operating expenses (EBIT).
    pub operating_income_usd: f64,
    /// Operating income − interest.
    pub pretax_income_usd: f64,
    /// Tax on pre-tax income (0 when pre-tax income is not positive).
    pub income_tax_usd: f64,
    /// Pre-tax income − tax.
    pub net_income_usd: f64,
    pub gross_margin_pct: f64,
    pub operating_margin_pct: f64,
    pub pretax_margin_pct: f64,
    pub net_margin_pct: f64,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &IncomeStatementInput) -> IncomeStatement {
    let gross = cents(i.revenue_usd - i.cogs_usd);
    let ebit = cents(gross - i.operating_expenses_usd);
    let pretax = cents(ebit - i.interest_expense_usd);
    // Tax applies only to positive pre-tax income.
    let tax = if pretax > 0.0 {
        cents(pretax * i.tax_rate_pct / 100.0)
    } else {
        0.0
    };
    let net = cents(pretax - tax);

    let margin = |x: f64| -> f64 {
        if i.revenue_usd != 0.0 {
            cents(x / i.revenue_usd * 100.0)
        } else {
            0.0
        }
    };
    let gm = margin(gross);
    let om = margin(ebit);
    let pm = margin(pretax);
    let nm = margin(net);

    let statement_body = format!(
        "Revenue {}\n  − Cost of goods sold {}\n  = Gross profit {} ({:.2}% margin)\n  − Operating expenses {}\n  = Operating income (EBIT) {} ({:.2}% margin)\n  − Interest expense {}\n  = Pre-tax income {} ({:.2}% margin)\n  − Income tax ({:.2}%) {}\n  = Net income {} ({:.2}% margin)",
        money(i.revenue_usd),
        money(i.cogs_usd),
        money(gross), gm,
        money(i.operating_expenses_usd),
        money(ebit), om,
        money(i.interest_expense_usd),
        money(pretax), pm,
        i.tax_rate_pct, money(tax),
        money(net), nm
    );

    let clauses = vec![
        DocClause {
            heading: "Statement".into(),
            body: format!("Company: {}\nPeriod: {}\nDate: {}", i.company_name, i.period_label, i.date),
        },
        DocClause { heading: "Income Statement".into(), body: statement_body },
        DocClause {
            heading: "Summary".into(),
            body: format!(
                "Gross profit {}, operating income {}, pre-tax income {}, net income {}.",
                money(gross), money(ebit), money(pretax), money(net)
            ),
        },
    ];

    IncomeStatement {
        title: "Income Statement (P&L)".into(),
        gross_profit_usd: gross,
        operating_income_usd: ebit,
        pretax_income_usd: pretax,
        income_tax_usd: tax,
        net_income_usd: net,
        gross_margin_pct: gm,
        operating_margin_pct: om,
        pretax_margin_pct: pm,
        net_margin_pct: nm,
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> IncomeStatementInput {
        IncomeStatementInput {
            company_name: "Acme Co".into(),
            period_label: "FY2025".into(),
            revenue_usd: 1_000_000.0,
            cogs_usd: 600_000.0,
            operating_expenses_usd: 250_000.0,
            interest_expense_usd: 20_000.0,
            tax_rate_pct: 21.0,
            date: "2026-01-31".into(),
        }
    }

    #[test]
    fn cascade_and_margins() {
        let d = generate(&base());
        assert!(close(d.gross_profit_usd, 400_000.0));
        assert!(close(d.operating_income_usd, 150_000.0));
        assert!(close(d.pretax_income_usd, 130_000.0));
        assert!(close(d.income_tax_usd, 27_300.0));
        assert!(close(d.net_income_usd, 102_700.0));
        assert!(close(d.gross_margin_pct, 40.0));
        assert!(close(d.operating_margin_pct, 15.0));
        assert!(close(d.pretax_margin_pct, 13.0));
        assert!(close(d.net_margin_pct, 10.27));
    }

    #[test]
    fn no_tax_on_pretax_loss() {
        let d = generate(&IncomeStatementInput { operating_expenses_usd: 500_000.0, ..base() });
        // gross 400k − opex 500k = −100k EBIT; pretax −120k → no tax.
        assert!(close(d.pretax_income_usd, -120_000.0));
        assert!(close(d.income_tax_usd, 0.0));
        assert!(close(d.net_income_usd, -120_000.0));
    }

    #[test]
    fn net_equals_pretax_minus_tax() {
        let d = generate(&base());
        assert!(close(d.net_income_usd, d.pretax_income_usd - d.income_tax_usd));
    }

    #[test]
    fn zero_revenue_zero_margins() {
        let d = generate(&IncomeStatementInput { revenue_usd: 0.0, cogs_usd: 0.0, operating_expenses_usd: 0.0, interest_expense_usd: 0.0, ..base() });
        assert!(close(d.gross_margin_pct, 0.0));
        assert!(close(d.net_income_usd, 0.0));
    }

    #[test]
    fn statement_lists_all_levels() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading == "Income Statement").unwrap();
        assert!(c.body.contains("Gross profit"));
        assert!(c.body.contains("Operating income"));
        assert!(c.body.contains("Net income"));
    }
}
